File: mcp-server/src/tools/categorizeTransactions.ts
import { z } from 'zod';
import { apiPost } from '../utils/apiClient.js';

export const categorizeTransactionsTool = {
  name: 'categorize_transactions' as const,
  description: 'Batch update the categories of transactions. Provide an array of objects with transaction id and new category. Common categories: groceries, dining, transportation, utilities, rent, salary, entertainment, healthcare, shopping, subscriptions, travel, freelance, interest.',
  parameters: z.object({
    updates: z.array(z.object({
      id: z.number().describe('Transaction ID'),
      category: z.string().describe('New category for the transaction'),
    })).describe('Array of transaction updates'),
  }),
  execute: async (args: { updates: Array<{ id: number; category: string }> }) => {
    const data = await apiPost('/api/transactions/categorize', { updates: args.updates });
    return JSON.stringify(data);
  },
};

File: mcp-server/src/tools/categorize_transactions.py
import json
from typing import Annotated

from fastmcp import FastMCP
from pydantic import Field

from ..utils.api_client import api_post
from ..utils.auth import get_current_token


def register(mcp: FastMCP) -> None:
    @mcp.tool(
        description="Batch update the categories of transactions. Provide an array of "
        "objects with transaction id and new category. Common categories: groceries, "
        "dining, transportation, utilities, rent, salary, entertainment, healthcare, "
        "shopping, subscriptions, travel, freelance, interest.",
    )
    async def categorize_transactions(
        updates: Annotated[
            list[dict[str, int | str]],
            Field(
                description="Array of transaction updates, each with 'id' (number) and 'category' (string)",
            ),
        ],
    ) -> str:
        token = get_current_token()
        data = await api_post("/api/transactions/categorize", {"updates": updates}, token)
        return json.dumps(data)
